KOGITO-9691 Create SonataFlow operator BDD tests for devMode

MarianMacik opened the pull request:
**JIRA:** [KOGITO-9691](https://issues.redhat.com/browse/KOGITO-9691)

This is a rewrite of the devMode end-to-end test to BDD test framework. Currently it lacks 2 things:

* OpenShift integration hasn't been tested yet
* Automatic execution in the PR checks, this will also provide a separate Makefile for easier execution

Those will be added shortly, not necessarily in this PR, but we will see.

<details>
<summary>
How to backport a pull request to a different branch?
</summary>

In order to automatically create a **backporting pull request** please add one or more labels having the following format `backport-<branch-name>`, where `<branch-name>` is the name of the branch where the pull request must be backported to (e.g., `backport-7.67.x` to backport the original PR to the `7.67.x` branch).

> **NOTE**: **backporting** is an action aiming to move a change (usually a commit) from a branch (usually the main one) to another one, which is generally referring to a still maintained release branch. Keeping it simple: it is about to move a specific change or a set of them from one branch to another.

Once the original pull request is successfully merged, the automated action will create one backporting pull request per each label (with the previous format) that has been added.

If something goes wrong, the author will be notified and at this point a manual backporting is needed.

> **NOTE**: this automated backporting is triggered whenever a pull request on `main` branch is labeled or closed, but both conditions must be satisfied to get the new PR created.
</details>

ricardozanini commented on testbdd/README.md:
```diff
@@ -0,0 +1,20 @@
+# BDD tests
+
+Tests in this module are a rewrite and enhancement of the end-to-end tests available in the `/test` directory.
```
I think it's worth replacing the `test` directory with the BDD one. Also, do we really need it to be in a new exported go module? Where are we going to use it but here?

ricardozanini replied:
It makes sense, but I'd like to avoid importing kogito-operator here. I think we agreed on creating a new repo for this and refactoring here and there to reuse the common modules.

radtriste replied:
yes but we are working by steps there :)

ricardozanini commented on testbdd/README.md:
```diff
@@ -0,0 +1,20 @@
+Tests in this module are a rewrite and enhancement of the end-to-end tests available in the `/test` directory.
+They use [Godog](https://github.com/cucumber/godog) framework which is the official [Cucumber](https://cucumber.io/) BDD framework for Go and use [Gherkin](https://cucumber.io/docs/gherkin/) for writing test scenarios.
+
+Tests also make use of the [kogito-operator](https://github.com/kiegroup/kogito-operator/tree/main/test) framework for Kubernetes which simplifies communication with a Kubernetes cluster.
```
Can't we bring it here? Not sure if I like the idea of importing this module.

radtriste replied:
you should extract the bdd test framework in a separate repository. and reuse in both operator repositories

ricardozanini reviewed (commented):
My only concerns are related to this test framework. We can either have it exported from `kogito-operator` to a separated go module or create a new git repo just for it. I think it's worth it, honestly. It will be a great addition to the community.

radtriste commented on testbdd/features/e2e.feature:
```diff
@@ -0,0 +1,24 @@
+  Scenario: order-processing E2E test
+    Given Namespace is created
+    When SonataFlow Operator is deployed
+    When SonataFlowPlatform is deployed
```
```suggestion
    When SonataFlowPlatform is created
```
wdyt  ?

radtriste commented on testbdd/features/e2e.feature:
```diff
@@ -0,0 +1,24 @@
+    When SonataFlow orderprocessing example is deployed
+    Then SonataFlow "orderprocessing" has the condition "Running" set to "True" within 1 minutes
+    Then SonataFlow "orderprocessing" is addressable within 1 minute
+    Then HTTP POST request as Cloud Event on SonataFlow "orderprocessing" is successful within 1 minute with path "", headers "content-type= application/json,ce-specversion= 1.0,ce-source= /from/localhost,ce-type= orderEvent,ce-id= f0643c68-609c-48aa-a820-5df423fa4fe0" and body:
+    """json
+    {"id":"f0643c68-609c-48aa-a820-5df423fa4fe0",
+    "country":"Czech Republic",
+    "total":10000,
+    "description":"iPhone 12"
+    }
+    """
```
I was thinking about maybe having all of this (path, headers, body) in a datatable instead. wdyt ?

but maybe could be for an imporvement of the framework

radtriste commented on testbdd/steps/sonataflowplatform.go:
```diff
@@ -0,0 +1,64 @@
+	projectDir, _ := utils.GetProjectDir()
+	projectDir = strings.Replace(projectDir, "/testbdd", "", -1)
+
+	// TODO or kubectl
```
yep. we should be able to run with kube only

radtriste reviewed (approved):
Overall all good for me for a first implementation.
We could rework the details later.

MarianMacik commented:
@radtriste @ricardozanini 
I have exported the framework including the whole Git history into a separate module in the `/bddframework` directory. History is relevant just for the framework and some other parts, which I included in it.

 What I did was:
* Added BDD Framework.
* Added kogito-operator kubernetes client which you @ricardozanini did for the Kogito operator. The framework heavily uses it so I decided to make it a part of the framework.
* I decided to retain some types like KogitoJobsService or KogitoDataIndex as they may be easily reused in the future should we need them.
* There is no longer a dependency on the kogito-operator repository.
* Added relevant hack scripts from the kogito-operator repo which are used by the tests.
* Added separate Makefiles for both the BDD tests (`testbdd`) and BDD Framework (`bddframework`).

What is remaining is to update the documentation and include this in a PR check. I have tested the scenario on Minikube and OpenShift already.

Thanks for the attention!

ricardozanini commented:
Marking it as unread, I'll carefully review it tomorrow. It's a waaay long PR :D

ricardozanini commented on bddframework/pkg/api/app/v1beta1/groupversion_info.go:
```diff
@@ -0,0 +1,42 @@
+
+var (
+	// GroupVersion is group version used to register these objects
+	GroupVersion = schema.GroupVersion{Group: "app.kiegroup.org", Version: "v1beta1"}
```
These types were migrated to keep compatibility with the previous work, but are not used here, right? This version follows the Kogito Operator.

MarianMacik replied:
Well, my initial thought was that if in future we are going to use some kind of supporting services like we used to in Kogito (be it Data Index, UI or whatever), this may be a good starting point in tests to reuse those steps and just repurpose these structs. But yes, those are project-specific, so before extracting the whole framework to a separate repository we could first move them under the `testbdd` module or simply remove them.

ricardozanini commented on bddframework/pkg/api/app/v1beta1/kogitoservices.go:
```diff
@@ -0,0 +1,447 @@
+)
+
+// KogitoServiceStatus is the basic structure for any Kogito Service status.
+type KogitoServiceStatus struct {
```
Assuming this is a compatibility migration only, we won't use this struct at all.

ricardozanini reviewed (approved):
Seems great! But not sure if we need this Kogito Operator's types. Worry not though, we can remove them later.
